src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -107,644 +93,557 @@ public void Emit()
 
-                StringBuilder sb = new("""
-// <auto-generated/>
+                    // Suppress warnings about [Obsolete] member usage in generated code.
+                    #pragma warning disable CS0612, CS0618
 
-#nullable enable annotations
-#nullable disable warnings
-
-// Suppress warnings about [Obsolete] member usage in generated code.
-#pragma warning disable CS0612, CS0618
+                    """);
 
+                if (contextSpec.Namespace != JsonConstants.GlobalNamespaceValue)
+                {
+                    writer.WriteLine($"namespace {contextSpec.Namespace}");
+                    writer.WriteLine('{');
+                    writer.Indentation++;
+                }
 
-""");
-                int indentation = 0;
+                ImmutableEquatableArray<string> contextClasses = contextSpec.ContextClassDeclarations;
+                Debug.Assert(contextClasses.Count > 0);
 
-                if (!isInGlobalNamespace)
+                // Emit any containing classes first.
+                for (int i = contextClasses.Count - 1; i > 0; i--)
                 {
-                    sb.AppendLine($$"""
-namespace {{@namespace}}
-{
-""");
-                    indentation++;
+                    writer.WriteLine(contextClasses[i]);
+                    writer.WriteLine('{');
+                    writer.Indentation++;
                 }
 
-                for (int i = 0; i < declarationCount - 1; i++)
+                if (isPrimaryContextSourceFile)
                 {
-                    string declarationSource = $$"""
-{{declarationList[declarationCount - 1 - i]}}
-{
-""";
-                    sb.AppendLine(IndentSource(declarationSource, numIndentations: indentation++));
+                    // Annotate context class with the GeneratedCodeAttribute
+                    writer.WriteLine($"""[global::System.CodeDom.Compiler.GeneratedCodeAttribute("{s_assemblyName.Name}", "{s_assemblyName.Version}")]""");
                 }
 
-                // Add the core implementation for the derived context class.
-                string partialContextImplementation = $$"""
-{{generatedCodeAttributeSource}}{{declarationList[0]}}{{(interfaceImplementation is null ? "" : ": " + interfaceImplementation)}}
-{
-{{IndentSource(source, 1)}}
-}
-""";
-                sb.AppendLine(IndentSource(partialContextImplementation, numIndentations: indentation--));
+                // Emit the JsonSerializerContext class declaration
+                writer.WriteLine($"{contextClasses[0]}{(interfaceImplementation is null ? "" : " : " + interfaceImplementation)}");
+                writer.WriteLine('{');
+                writer.Indentation++;
 
-                // Match curly braces for each containing type/namespace.
-                for (; indentation >= 0; indentation--)
+                return writer;
+            }
+
+            private static SourceText CompleteSourceFileAndReturnText(SourceWriter writer)
```

Comment:
Could this be a helper on the writer?

Reply:
It could, although I wanted to make the writer language agnostic (and only concert itself with the responsibility of track indentation).